<!DOCTYPE html>
<html lang="zh-cn" dir="ltr">
    <head><meta charset='utf-8'>
<meta name='viewport' content='width=device-width, initial-scale=1'><meta name='description' content="Bulk-RNA测序上游分析——从reads到counts 转录组基础知识 以下记载了转录组最基础的一些知识，这些概念在RNA测序分析中是至关重要的：\n1、基因组Genome：指某一特定物种细胞内或病毒颗粒内的一整套遗传物质（DNA或RNA）。对于人类而言，人类细胞是二倍体，有2套染色体，其中任何一套染色体的DNA序列即为一个基因组，比如人类基因组就是1-22号染色体+X+Y。在广义上，线粒体DNA也可以算作基因组的一部分。\n">
<title>Rna测序分析的标准流程</title>

<link rel='canonical' href='https://Detarame-ding.github.io/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/'>

<link rel="stylesheet" href="/scss/style.min.663803bebe609202d5b39d848f2d7c2dc8b598a2d879efa079fa88893d29c49c.css"><meta property='og:title' content="Rna测序分析的标准流程">
<meta property='og:description' content="Bulk-RNA测序上游分析——从reads到counts 转录组基础知识 以下记载了转录组最基础的一些知识，这些概念在RNA测序分析中是至关重要的：\n1、基因组Genome：指某一特定物种细胞内或病毒颗粒内的一整套遗传物质（DNA或RNA）。对于人类而言，人类细胞是二倍体，有2套染色体，其中任何一套染色体的DNA序列即为一个基因组，比如人类基因组就是1-22号染色体+X+Y。在广义上，线粒体DNA也可以算作基因组的一部分。\n">
<meta property='og:url' content='https://Detarame-ding.github.io/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/'>
<meta property='og:site_name' content='鳕鱼-Detarame'>
<meta property='og:type' content='article'><meta property='article:section' content='Post' /><meta property='article:published_time' content='2024-12-14T00:00:00&#43;00:00'/><meta property='article:modified_time' content='2024-12-14T00:00:00&#43;00:00'/>
<meta name="twitter:title" content="Rna测序分析的标准流程">
<meta name="twitter:description" content="Bulk-RNA测序上游分析——从reads到counts 转录组基础知识 以下记载了转录组最基础的一些知识，这些概念在RNA测序分析中是至关重要的：\n1、基因组Genome：指某一特定物种细胞内或病毒颗粒内的一整套遗传物质（DNA或RNA）。对于人类而言，人类细胞是二倍体，有2套染色体，其中任何一套染色体的DNA序列即为一个基因组，比如人类基因组就是1-22号染色体+X+Y。在广义上，线粒体DNA也可以算作基因组的一部分。\n">
    <link rel="shortcut icon" href="/favicon.ico" />

    </head>
    <body class="
    article-page
    ">
    <script>
        (function() {
            const colorSchemeKey = 'StackColorScheme';
            if(!localStorage.getItem(colorSchemeKey)){
                localStorage.setItem(colorSchemeKey, "auto");
            }
        })();
    </script><script>
    (function() {
        const colorSchemeKey = 'StackColorScheme';
        const colorSchemeItem = localStorage.getItem(colorSchemeKey);
        const supportDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches === true;

        if (colorSchemeItem == 'dark' || colorSchemeItem === 'auto' && supportDarkMode) {
            

            document.documentElement.dataset.scheme = 'dark';
        } else {
            document.documentElement.dataset.scheme = 'light';
        }
    })();
</script>
<div class="container main-container flex on-phone--column extended"><aside class="sidebar left-sidebar sticky ">
    <button class="hamburger hamburger--spin" type="button" id="toggle-menu" aria-label="切换菜单">
        <span class="hamburger-box">
            <span class="hamburger-inner"></span>
        </span>
    </button>

    <header>
        
            
            <figure class="site-avatar">
                <a href="/">
                
                    
                    
                    
                        
                        <img src="/img/OIP_hu8127449034170590479.jpg" width="300"
                            height="298" class="site-logo" loading="lazy" alt="Avatar">
                    
                
                </a>
                
                    <span class="emoji">😊</span>
                
            </figure>
            
        
        
        <div class="site-meta">
            <h1 class="site-name"><a href="/">鳕鱼-Detarame</a></h1>
            <h2 class="site-description"></h2>
        </div>
    </header><ol class="menu-social">
            
                <li>
                    <a 
                        href='https://github.com/CaiJimmy/hugo-theme-stack'
                        target="_blank"
                        title="GitHub"
                        rel="me"
                    >
                        
                        
                            <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-brand-github" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
  <path d="M9 19c-4.3 1.4 -4.3 -2.5 -6 -3m12 5v-3.5c0 -1 .1 -1.4 -.5 -2c2.8 -.3 5.5 -1.4 5.5 -6a4.6 4.6 0 0 0 -1.3 -3.2a4.2 4.2 0 0 0 -.1 -3.2s-1.1 -.3 -3.5 1.3a12.3 12.3 0 0 0 -6.2 0c-2.4 -1.6 -3.5 -1.3 -3.5 -1.3a4.2 4.2 0 0 0 -.1 3.2a4.6 4.6 0 0 0 -1.3 3.2c0 4.6 2.7 5.7 5.5 6c-.6 .6 -.6 1.2 -.5 2v3.5" />
</svg>



                        
                    </a>
                </li>
            
                <li>
                    <a 
                        href='https://twitter.com'
                        target="_blank"
                        title="Twitter"
                        rel="me"
                    >
                        
                        
                            <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-brand-twitter" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
  <path d="M22 4.01c-1 .49 -1.98 .689 -3 .99c-1.121 -1.265 -2.783 -1.335 -4.38 -.737s-2.643 2.06 -2.62 3.737v1c-3.245 .083 -6.135 -1.395 -8 -4c0 0 -4.182 7.433 4 11c-1.872 1.247 -3.739 2.088 -6 2c3.308 1.803 6.913 2.423 10.034 1.517c3.58 -1.04 6.522 -3.723 7.651 -7.742a13.84 13.84 0 0 0 .497 -3.753c-.002 -.249 1.51 -2.772 1.818 -4.013z" />
</svg>



                        
                    </a>
                </li>
            
        </ol><ol class="menu" id="main-menu">
        
        
        
        <li >
            <a href='/' >
                
                
                
                    <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-home" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <polyline points="5 12 3 12 12 3 21 12 19 12" />
  <path d="M5 12v7a2 2 0 0 0 2 2h10a2 2 0 0 0 2 -2v-7" />
  <path d="M9 21v-6a2 2 0 0 1 2 -2h2a2 2 0 0 1 2 2v6" />
</svg>



                
                <span>主页</span>
            </a>
        </li>
        
        
        <li >
            <a href='/%E5%85%B3%E4%BA%8E/' >
                
                
                
                    <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-user" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="12" cy="7" r="4" />
  <path d="M6 21v-2a4 4 0 0 1 4 -4h4a4 4 0 0 1 4 4v2" />
</svg>



                
                <span>关于</span>
            </a>
        </li>
        
        
        <li >
            <a href='/archives/' >
                
                
                
                    <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-archive" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <rect x="3" y="4" width="18" height="4" rx="2" />
  <path d="M5 8v10a2 2 0 0 0 2 2h10a2 2 0 0 0 2 -2v-10" />
  <line x1="10" y1="12" x2="14" y2="12" />
</svg>



                
                <span>Archives</span>
            </a>
        </li>
        
        
        <li >
            <a href='/search/' >
                
                
                
                    <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-search" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="10" cy="10" r="7" />
  <line x1="21" y1="21" x2="15" y2="15" />
</svg>



                
                <span>Search</span>
            </a>
        </li>
        
        
        <li >
            <a href='/links/' >
                
                
                
                    <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-link" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <path d="M10 14a3.5 3.5 0 0 0 5 0l4 -4a3.5 3.5 0 0 0 -5 -5l-.5 .5" />
  <path d="M14 10a3.5 3.5 0 0 0 -5 0l-4 4a3.5 3.5 0 0 0 5 5l.5 -.5" />
</svg>



                
                <span>Links</span>
            </a>
        </li>
        
        <li class="menu-bottom-section">
            <ol class="menu">

                
                    <li id="dark-mode-toggle">
                        <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-toggle-left" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="8" cy="12" r="2" />
  <rect x="2" y="6" width="20" height="12" rx="6" />
</svg>



                        <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-toggle-right" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="16" cy="12" r="2" />
  <rect x="2" y="6" width="20" height="12" rx="6" />
</svg>



                        <span>暗色模式</span>
                    </li>
                
            </ol>
        </li>
    </ol>
</aside>

    <aside class="sidebar right-sidebar sticky">
        
            
                
    <section class="widget archives">
        <div class="widget-icon">
            <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-hash" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <line x1="5" y1="9" x2="19" y2="9" />
  <line x1="5" y1="15" x2="19" y2="15" />
  <line x1="11" y1="4" x2="7" y2="20" />
  <line x1="17" y1="4" x2="13" y2="20" />
</svg>



        </div>
        <h2 class="widget-title section-title">目录</h2>
        
        <div class="widget--toc">
            <nav id="TableOfContents">
  <ol>
    <li><a href="#bulk-rna测序上游分析从reads到counts">Bulk-RNA测序上游分析——从reads到counts</a>
      <ol>
        <li><a href="#转录组基础知识">转录组基础知识</a></li>
        <li><a href="#硬件和软件的准备">硬件和软件的准备</a></li>
        <li><a href="#fastq和fasta文件概述">FASTQ和FASTA文件概述</a></li>
        <li><a href="#质量控制与预处理">质量控制与预处理</a></li>
        <li><a href="#将read比对到参考基因组">将read比对到参考基因组</a>
          <ol>
            <li><a href="#讨论">讨论</a></li>
          </ol>
        </li>
        <li><a href="#定量和基于注释的质量控制">定量和基于注释的质量控制</a>
          <ol>
            <li><a href="#讨论-1">讨论</a></li>
          </ol>
        </li>
        <li><a href="#bulk-rna测序下游分析从counts开始">Bulk-RNA测序下游分析——从counts开始</a>
          <ol>
            <li><a href="#数据准备">数据准备</a></li>
            <li><a href="#使用r语言进行分析">使用R语言进行分析</a></li>
          </ol>
        </li>
      </ol>
    </li>
  </ol>
</nav>
        </div>
    </section>

            
        
    </aside>


            <main class="main full-width">
    <article class="main-article">
    <header class="article-header">

    <div class="article-details">
    
    <header class="article-category">
        
            <a href="/categories/rnaseq/" >
                RnaSeq
            </a>
        
    </header>
    

    <div class="article-title-wrapper">
        <h2 class="article-title">
            <a href="/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/">Rna测序分析的标准流程</a>
        </h2>
    
        
    </div>

    
    
    
    
    <footer class="article-time">
        
            <div>
                <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-calendar-time" width="56" height="56" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <path d="M11.795 21h-6.795a2 2 0 0 1 -2 -2v-12a2 2 0 0 1 2 -2h12a2 2 0 0 1 2 2v4" />
  <circle cx="18" cy="18" r="4" />
  <path d="M15 3v4" />
  <path d="M7 3v4" />
  <path d="M3 11h16" />
  <path d="M18 16.496v1.504l1 1" />
</svg>
                <time class="article-time--published">2024-12-14</time>
            </div>
        

        
            <div>
                <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-clock" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="12" cy="12" r="9" />
  <polyline points="12 7 12 12 15 15" />
</svg>



                <time class="article-time--reading">
                    阅读时长: 18 分钟
                </time>
            </div>
        
    </footer>
    

    
</div>

</header>

    <section class="article-content">
    
    
    <h2 id="bulk-rna测序上游分析从reads到counts">Bulk-RNA测序上游分析——从reads到counts
</h2><h3 id="转录组基础知识">转录组基础知识
</h3><p>以下记载了转录组最基础的一些知识，这些概念在RNA测序分析中是至关重要的：</p>
<p>1、基因组Genome：指某一特定物种细胞内或病毒颗粒内的一整套遗传物质（DNA或RNA）。对于人类而言，人类细胞是二倍体，有2套染色体，其中任何一套染色体的DNA序列即为一个基因组，比如人类基因组就是1-22号染色体+X+Y。在广义上，线粒体DNA也可以算作基因组的一部分。</p>
<p>2、转录组Transcriptome：广义上指在某种条件下在一个细胞或一群细胞中所转录出的RNA总和，包括mRNA、rRNA、tRNA和其他非编码RNA等。狭义的转录组特指转录出的mRNA。转录组和基因组的差别在于，基因组在给定的细胞株上一般不变，而转录组可能会有很大变化，它反映了在给定条件下细胞内活跃表达的基因，故也成为“基因表达谱”。</p>
<p>3、转录Transcription：即在RNA聚合酶的作用下，遗传信息由DNA复制到RNA的过程。其分为三个过程：启动、延伸和终止。由RNA聚合酶和转录因子共同结合到模板DNA的启动子序列上开启转录过程，接着按互补配对原则进行延伸，最后通过依赖ρ因子的终止或非依赖ρ因子的终止来终止转录。转录出的初级转录RNA包括以下区域：5’UTR（5’非翻译区）、3’UTR（3’非翻译区）、Intron（内含子）、Exon（外显子）。</p>
<p>4、转录后修饰Post-transcriptional modification：是指在真核细胞中，将初级转录RNA转化为成熟RNA的过程。包括3个过程：5’端加帽、3’端加尾以及RNA剪切。</p>
<p>5、可变剪接Alternative splicing：又称可变剪接，即一条未经剪接的RNA，含有的多种外显子被剪成不同的组合，因而翻译成不同的蛋白质。剪切出的不同mRNA即为mRNA异构体（isomer）。</p>
<p>6、转录本Transcript：即一条mRNA。一条基因可能有不同的转录本，其中一个原因就是可变剪切。</p>
<p><img src="/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/RNAstructure.png"
	width="1688"
	height="465"
	srcset="/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/RNAstructure_hu17240095561173243829.png 480w, /p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/RNAstructure_hu14383290478521594772.png 1024w"
	loading="lazy"
	
		alt="RNAstructure"
	
	
		class="gallery-image" 
		data-flex-grow="363"
		data-flex-basis="871px"
	
></p>
<h3 id="硬件和软件的准备">硬件和软件的准备
</h3><p>1、个人计算机的准备:</p>
<ul>
<li>
<p>操作系统：Ubuntu 24.04 LTS</p>
</li>
<li>
<p>RAM：16G；硬盘：500G</p>
</li>
</ul>
<p>2、软件的准备。</p>
<ul>
<li>
<p>不具体介绍某个软件的下载方法</p>
</li>
<li>
<p>一般的软件安装命令方法如下：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">sudo apt install Name_of_Software
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>或者直接在Github等网站上下载</p>
</li>
</ul>
<p>3、将软件所在位置添加入环境变量，或将PYTHON脚本添加到PYTHON环境变量:</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl"><span class="nb">echo</span> <span class="s1">&#39;export PATH=path/to/your/software:$PATH&#39;</span> &gt;&gt; ~/.bashrc
</span></span><span class="line"><span class="cl"><span class="nb">echo</span> <span class="s1">&#39;export PYTHONPATH=path/to/your/python_script:$PYTHONPATH&#39;</span> &gt;&gt; ~/.bashrc
</span></span><span class="line"><span class="cl"><span class="nb">source</span> ~/.bashrc <span class="c1">#重新加载环境变量</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="fastq和fasta文件概述">FASTQ和FASTA文件概述
</h3><p>RNA测序分析的第一步是从测序的原始数据开始的。测序公司返回的原始数据文件是FASTQ格式，里面的每一条记录就是一个read读段。还有一种类似的数据格式叫做FASTA，一般用来记录已经组装好的、参考基因组的序列。</p>
<p>使用以下命令下载本例所需的文件：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqGm12892R2x75Il200FastqRd1Rep2V2.fastq.gz
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqGm12892R2x75Il200FastqRd2Rep2V2.fastq.gz
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl">wget ftp://ftp.ensembl.org/pub/release-74/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.74.dna.toplevel.fa.gz
</span></span></code></pre></td></tr></table>
</div>
</div><p>其中，前两个文件为来自淋巴母细胞系的双端测序文件，数据在加利福尼亚理工学院产生，来自早期Illumina平台。下载完毕后解压并重命名为Read1.fastq，Read2.fastq。第三个文件是从ENSRMBL网站上下载的人类参考基因组，解压后大小约为30Gb。下载完解压并重命名为Homo_sapiens.GRCh37.74.dna.toplevel.fa。</p>
<p>分别打开Read1.fastq和Read2.fastq两个文件，可以发现其前四行分别为：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">@HWUSI-EAS1665_28:5:1:1038:939/1
</span></span><span class="line"><span class="cl">NNNNGGGGTAGGAGNNNNNNNNNGGAGCCAAGGGGGCGTGGCTACGATCTGTGGGACTATGACTGAAATGCTGTA
</span></span><span class="line"><span class="cl">+
</span></span><span class="line"><span class="cl">BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
</span></span></code></pre></td></tr></table>
</div>
</div><p>以及</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">@HWUSI-EAS1665_28:5:1:1038:939/2
</span></span><span class="line"><span class="cl">GCACAGGCGGGGGTCACTCTTTTTTGGGTCCTCCGAAATTCGCGGGGAGGATTCAACATCACGAAGCTTGCCGCA
</span></span><span class="line"><span class="cl">+
</span></span><span class="line"><span class="cl">BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
</span></span></code></pre></td></tr></table>
</div>
</div><p>Read1.fastq和Read2.fastq中每一条记录都是一一对应的，代表同一条mRNA的两端的序列。每一条fastq记录都由4行组成：</p>
<ul>
<li>
<p>第一行记录了测序时的相关信息，其中@HWUSI-EAS1665_28是测序机器名称，5代表测序槽lane的序号，1代表测序小室tile的编号，1038和939分别是tile内的XY坐标，/1代表是正向测序序列，/2代表是逆向测序序列。</p>
</li>
<li>
<p>第二行是测得的mRNA序列，本里中所有的记录都是75bp的，用ATCG代表四种碱基，N代表不确定的模糊碱基。</p>
</li>
<li>
<p>第三行是一个加号（+）作为分隔符。</p>
</li>
<li>
<p>第四行是碱基的质量Q，Q为-10*lg(P)再四舍五入到整数，其中P是碱基错误的概率，再将Q加上33转化为ASCII码（phred33编码）【如果是低于1.8版本的Illumina软件产生的FASTQ文件则是Q+64转化为ASCII码（phred64编码），本例就是phred64编码的】。</p>
</li>
</ul>
<p>接着，打开Homo_sapiens.GRCh37.74.dna.toplevel.fa，其前两行为：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">&gt;1 dna:chromosome chromosome:GRCh37:1:1:249250621:1 REF
</span></span><span class="line"><span class="cl">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN…
</span></span></code></pre></td></tr></table>
</div>
</div><p>该文件的每条记录都由2行组成：</p>
<ul>
<li>
<p>第一行记录了基因组的相关信息，本例中&gt;1代表是1号染色体，dna:chromosome代表记录的是染色体DNA，chromosome:GRCh37代表是人类基因组版本GRCh37的数据，1代表染色体编号为1，1:249250621代表起始位置从1到249250621，1代表正链，REF说明这是个参考序列。</p>
</li>
<li>
<p>第二行即为基因序列。</p>
</li>
</ul>
<h3 id="质量控制与预处理">质量控制与预处理
</h3><p>1、用fastqc对原始文件生成一个质量报告。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">fastqc Read1.fastq.gz
</span></span><span class="line"><span class="cl">fastqc Read2.fastq.gz
</span></span></code></pre></td></tr></table>
</div>
</div><p>2、使用Trimmomatic【Trimmomatic版本0.39】过滤掉平均质量低于20（AVGQUAL:20）的read，采用双端测序模式（PE）。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl"><span class="nb">cd</span> RnaSeq_Fastq_data <span class="c1">#注：下载的所有数据都是放在~/RnaSeq_Fastq_data文件夹下的，下不赘述</span>
</span></span><span class="line"><span class="cl">java -jar ~/software/Trimmomatic-0.39/trimmomatic-0.39.jar PE -phred64 Read1.fastq.gz Read2.fastq.gz paired1.fq.gz unpaired1.fq.gz paired2.fq.gz unpaired2.fq.gz AVGQUAL:20
</span></span></code></pre></td></tr></table>
</div>
</div><p>其中生成的upaired1/2.fq.gz文件存储了经过过滤幸存、但失去其配对的read。</p>
<p>3、使用Trimmomatic修剪，当碱基质量低于20(TRAILING:20)时从3’端修剪碱基，并且去除修剪后长度低于50bp(MINLEN:50)的read。该命令作用于双端read。其他修剪方法暂略。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl"><span class="nb">cd</span> RnaSeq_Fastq_data
</span></span><span class="line"><span class="cl">java -jar ~/software/Trimmomatic-0.39/trimmomatic-0.39.jar PE -phred64 Read1.fastq.gz Read2.fastq.gz paired1.fq.gz unpaired1.fq.gz paired2.fq.gz unpaired2.fq.gz TRAILING:20 MINLEN:50
</span></span></code></pre></td></tr></table>
</div>
</div><p>4、接头的去除。案例中原始数据已经进行接头的去除，这里暂略接头的去除方法。</p>
<p>5、使用prinseq-lite【prinseq-lite版本0.20.4】进行筛选。去除长度小于50的read（-min_len 50），去除平均质量低于20的read（-min_qual_mean 20）【这一工作在已经在1和2中完成，可以省略】；去除低复杂度的read，即熵&lt;70的read（lc_method entrophy lc_threshold 70）；去除模糊碱基N&gt;2的read（ns_max_n 2）。经过滤过后保留配对的文件存储在nfiltered_1/2.fastq中，经过滤过幸存、但失去其配对的read保存在nfiltered_1/2_single-tons.fastq中。这一步骤可能要计算数个小时。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">prinseq-lite -fastq Read1.fastq -fastq2 Read2.fastq -min_len <span class="m">50</span> -min_qual_mean <span class="m">20</span> lc_method entrophy lc_threshold <span class="m">70</span> ns_max_n <span class="m">2</span> -out_good nfiltered -log -verbose
</span></span></code></pre></td></tr></table>
</div>
</div><p>6、重复read的删除。在RNA-seq中，重复往往是对高度表达的转录本进行测序的自然结果（高表达导致测序重复数多），不建议删除重复。如有必要可以用prinseq-lite来处理，删除出现100次以上（-derep_min 101）的准确重复段（-derep 1）。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">prinseq-lite -fastq Read1.fastq -fastq2 Read2.fastq -derep <span class="m">1</span> -derep_min <span class="m">101</span> -log -verbose -out_good dupfiltered -out_bad null -no_qual_header
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="将read比对到参考基因组">将read比对到参考基因组
</h3><p>1、如前所述，在Ensembl上下载人类参考基因组：在http://ftp.ensembl.org/pub/release-74/fasta/homo_sapiens/dna上下载Homo_sapiens.GRCh37.74.dna.toplevel.fa.gz文件并解压，解压后的文件大小约为33Gb。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">wget ftp://ftp.ensembl.org/pub/release-74/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.74.dna.toplevel.fa.gz
</span></span></code></pre></td></tr></table>
</div>
</div><p>2、用bowtie2【bowtie2版本2.5.4】生成索引。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">bowtie2-build ~/RnaSeq_Fastq_data/Homo_sapiens.GRCh37.74.dna.toplevel.fa/lustre/scratch110/ensembl/amonida/rel74/fasta_dumping_e74/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.74.dna.chromosome.1.fa GRCh37.74
</span></span></code></pre></td></tr></table>
</div>
</div><p>该操作会创建6个bt2l文件，均以CRCh37.74作为索引基本名。该操作会花费数个小时。生成的6个文件是GRCh37.74.1.bt2l；GRCh37.74.2.bt2l；GRCh37.74.3.bt2l；GRCh37.74.4.bt2l；GRCh37.74.rev.1.bt2l；GRCh37.74.rev.2.bt2l。他们的基本名是CRCh37.74。</p>
<p>3、将read比对到基因组。这里参考的索引基本名是CRCh37.7（-x CRCh47.7），输入文件（-U Read1.fastq.gz）是fastq格式（-q），它使用phred+64编码（&ndash;phred64）。将结果（-S）输出到Read1aligned.sam的文件，这里面不包括未比对的read（&ndash;no-unal）。有8个处理器同时使用以加快对比速度（-p 8）。该操作会花费数个小时。</p>
<ul>
<li>
<p>如果是单端测序：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">bowtie2 -q --phred64 -p <span class="m">8</span> --no-unal -x GRCh37.74 -U Read1.fastq.gz -S Read1aligned.sam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>如果是双端测序：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">bowtie2 -q --phred64 -p <span class="m">8</span> --no-unal -x GRCh37.74 -1 Read1.fastq.gz -2 Read2.fastq.gz -S Readaligned.sam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul>
<p>#Bowtie2在进行比对时不需要原始的fasta文件，只需要用生成的6个索引文件即可完成比对。比对结束后Bowtie2按照SAM格式报告比对结果，本例中文件名为Readaligned.sam。</p>
<p>4、使用samtools【samtools版本1.19.2】对SAM文件进一步处理。</p>
<ul>
<li>
<p>将SAM格式转化为BAM格式以节省空间，并且方便后续的操作。本例中指定输入是SAM文件（-S），输出是BAM文件（-b），输出文件名是alignments.bam（-o alignments.bam）。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools view -bS -o alignments.bam Readaligned.sam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>在BAM中按染色体坐标进行排序，或者使用名称排序。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools sort -o alignments.sorted.bam alignments.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>给BAM文件编制索引。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools index alignments.sorted.bam
</span></span></code></pre></td></tr></table>
</div>
</div><p>这将会生成一个BAI索引文件alignments.sorted.bam.bai</p>
</li>
<li>
<p>指定一个染色体或染色体区域，制作对比的子集。本例提取18号染色体的比对制作比对的子集。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools view -b -o alignments.sorted.18.bam alignments.sorted.bam <span class="m">18</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>基于作图质量过滤比对，本例中保留作图质量高于30的比对。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools view -b -q <span class="m">30</span> -o alignments_MQmin30.bam alignments.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>查看统计项目，如多少个read正确配对，或多少个read与其配对read作图到不同染色体等。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools flagstat alignment.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>BAM文件索引添加“chr”，如将“1”变成“ch1”（如果需要）：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools view -H H1Rep1.18.bam <span class="p">|</span> sed -e <span class="s1">&#39;s/SN:\([0-9XY]\)/SN:chr**\1**/&#39;</span> -e <span class="s1">&#39;s/SN:MT/SN:chrM/&#39;</span> <span class="p">|</span> samtools reheader - H1Rep1.18.bam &gt; H1Rep1.18.chradded.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul>
<p>4、可视化比对的read。可以使用Intergrative Genomics Viewer（IGV）【IGV版本2.4.10】软件对BAM文件进行可视化。可视化过程需要两个文件，即alignments.sorted.bam，以及它的索引文件alignments.sorted.bam.bai。本例中java版本需要切换到java8才能运行IGV</p>
<h4 id="讨论">讨论
</h4><h5 id="讨论1为什么要制作参考基因组索引">讨论1：为什么要制作参考基因组索引？
</h5><p>将一条read（本例中为75bp）匹配到数以亿计bp的人类基因组并非一件容易的事，而匹配一个样本全部数十万条记录可能会花上数百年的时间。对参考基因组进行预处理，就如同给字典加上目录一样，在查询的时候只需要从目录定位大致位置，再在小范围内查找即可（而不是遍历整个字典）。为参考基因组制作索引可以大大减小比对的时间复杂度，从而优化查找时间。Bowtie2采用的是FM-index压缩技术，该技术基于Burrows-Wheeler Transform（BWT）和Suffix Array（后缀数组）算法，能够在占用较小内存的同时实现快速的子字符串匹配。此外，制作的索引包含了原始fasta文件的所有信息，故bowtie2在进行匹配时只需要用到索引即可，不需要原始fasta文件。在本例中：</p>
<ul>
<li>
<p>GRCh37.74.1.bt2l、GRCh37.74.2.bt2l包含了FM-index的信息；</p>
</li>
<li>
<p>GRCh37.74.3.bt2l、GRCh37.74.4.bt2l包含了额外的索引结构辅助快速定位；</p>
</li>
<li>
<p>GRCh37.74.rev.1.bt2l、GRCh37.74.rev.2.bt2l包含了反向互补序列的FM-index信息。</p>
</li>
</ul>
<p>此外，这些都是二进制文件，无法直接读取。</p>
<h5 id="讨论2sam文件格式">讨论2：SAM文件格式
</h5><p>SAM格式储存了bowtie2将read与参考基因组进行比对后的结果。SAM文件内容如下：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-fallback" data-lang="fallback"><span class="line"><span class="cl">@HDVN:1.5SO:unsortedGO:query
</span></span><span class="line"><span class="cl">@SQSN:1LN:249250621
</span></span><span class="line"><span class="cl">…
</span></span><span class="line"><span class="cl">@SQSN:GL000207.1LN:4262
</span></span><span class="line"><span class="cl">@PGID:bowtie2PN:bowtie2VN:2.5.4CL:&#34;~/software/bowtie2-2.5.4/bowtie2-align-l --wrapper basic-0 -q --phred64 -p 8 -x GRCh37.74 --passthrough -1 Read1.fastq.gz -2 Read2.fastq.gz&#34;
</span></span><span class="line"><span class="cl">HWUSI-EAS1665_28:5:1:1242:9501531762289734275M=762289730TACAGCTACAAGACCTTTCTTGAAAATCTTATTTAATTCTGAGCCCATATTTCACTTACCTTATTTAAAATAAAT92B
</span></span><span class="line"><span class="cl">??B=:B?;;B-69&gt;G=GFE=F@&lt;4@@F@B55B4B,&lt;=B4==&lt;=3@88@B84&lt;BB-:47B@?=32068=,--4AS:i:-3XN:i:0XM:i:1XO:i:0XG:i:0NM:i:1MD:Z:13A61YT:Z:UP
</span></span><span class="line"><span class="cl">…
</span></span></code></pre></td></tr></table>
</div>
</div><p>其中：</p>
<ul>
<li>
<p>@HD行为文件头信息，包括版本（VN）、排序方式（SO）和排序方式（GO）</p>
</li>
<li>
<p>@SQ行为参考序列信息，包括染色体名称（SN）和长度（LN）</p>
</li>
<li>
<p>@PG行为程序信息，包括程序ID（ID）、程序名（PN）、版本号(VN)和指令（CL）</p>
</li>
<li>
<p>接下来每行都记录了一条比对信息，由11个必须字段和若干可选字段组成，中间用Tab分隔：</p>
</li>
</ul>
<div class="table-wrapper"><table>
  <thead>
      <tr>
          <th>字段编号</th>
          <th>字段名称</th>
          <th>说明</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td>1</td>
          <td>QNAME</td>
          <td>序列名称，本例为HWUSI-EAS1665_28:5:1:1242:950</td>
      </tr>
      <tr>
          <td>2</td>
          <td>FLAG</td>
          <td>比对标志位，用于描述比对的特征和属性，这里153转化为2进制为10011001，代表8个参数</td>
      </tr>
      <tr>
          <td>3</td>
          <td>RNAME</td>
          <td>比对到的参考序列名称，这里1代表比对到1号染色体</td>
      </tr>
      <tr>
          <td>4</td>
          <td>POS</td>
          <td>对比的起始位置，这里之1号染色体的76228973位</td>
      </tr>
      <tr>
          <td>5</td>
          <td>MAPQ</td>
          <td>比对的可信度</td>
      </tr>
      <tr>
          <td>6</td>
          <td>CIGAR</td>
          <td>CIGAR字符串，这里75M指75个碱基完全匹配。不同字母所代表的含义不做展开</td>
      </tr>
      <tr>
          <td>7</td>
          <td>RNEXT</td>
          <td>指当前read的配对read比对到的参考序列名称。“=”代表二者对比到了同一个参考序列</td>
      </tr>
      <tr>
          <td>8</td>
          <td>PNEXT</td>
          <td>下一个比对的起始位置</td>
      </tr>
      <tr>
          <td>9</td>
          <td>TLEN</td>
          <td>模板长度，表示该read在参考序列上的覆盖长度。这里0代表数据不可用</td>
      </tr>
      <tr>
          <td>10</td>
          <td>SEQ</td>
          <td>read序列</td>
      </tr>
      <tr>
          <td>11</td>
          <td>QUAL</td>
          <td>read质量，phred64编码</td>
      </tr>
      <tr>
          <td>+1</td>
          <td>AS:i</td>
          <td>比对得分</td>
      </tr>
      <tr>
          <td>+2</td>
          <td>XN:i</td>
          <td>模糊匹配数，即匹配到多个位置的数量</td>
      </tr>
      <tr>
          <td>+3</td>
          <td>XM:i</td>
          <td>错配数</td>
      </tr>
      <tr>
          <td>+4</td>
          <td>XO:i</td>
          <td>比对过程中发生插入/删除（indel）事件的数量</td>
      </tr>
      <tr>
          <td>+5</td>
          <td>XG:i</td>
          <td>比对中发生插入或删除碱基的总数</td>
      </tr>
      <tr>
          <td>+6</td>
          <td>NM:i</td>
          <td>编辑距离</td>
      </tr>
      <tr>
          <td>+7</td>
          <td>MD:Z</td>
          <td>read和参考序列之间的错配位置与类型</td>
      </tr>
      <tr>
          <td>+8</td>
          <td>YT:Z</td>
          <td>比对类型</td>
      </tr>
  </tbody>
</table></div>
<h3 id="定量和基于注释的质量控制">定量和基于注释的质量控制
</h3><p>1、在UCSC Table Browser上获得人类基因组注释文件：在assembly菜单选择Feb. 2009 (GRCh37/hg19)；Group菜单选择Genes and gene predicitions；Track菜单选择Ensembl gene；region选择genome；output format选择BED – browser extensible data；output file写hg19_Ensembl_chr.bed（因为Ensembl数据集中染色体前缀包括“chr”）。</p>
<p>2、准备数据集</p>
<ul>
<li>
<p>在ensembl上下载18号染色体的fasta文件。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">wget -c https://ftp.ensembl.org/pub/release-74/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.74.dna.chromosome.18.fa.gz
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>使用wget命令下载wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep1V2.bam文件，它来自人类胚胎干细胞系，数据在加利福尼亚理工学院产生，来自早期Illumina平台。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep1V2.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>下载对应的注释文件。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-gdscript3" data-lang="gdscript3"><span class="line"><span class="cl"><span class="n">wget</span> <span class="o">-</span><span class="n">c</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">hgdownload</span><span class="o">.</span><span class="n">cse</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenPath</span><span class="o">/</span><span class="n">hg19</span><span class="o">/</span><span class="n">encodeDCC</span><span class="o">/</span><span class="n">wgEncodeCaltechRnaSeq</span><span class="o">/</span><span class="n">wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep1V2</span><span class="o">.</span><span class="n">bam</span><span class="o">.</span><span class="n">bai</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>为方便起见将文件分别重命名为H1Rep1.bam和H1Rep1.bam.bai</p>
</li>
<li>
<p>从H1Rep1.bam中提取比对到18号染色体的那一部分。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools view -b -o H1Rep1.18.bam H1Rep1.bam chr18
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>将H1Rep1.18.bam排序后制作索引</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">samtools sort -o H1Rep1.18.sorted.bam H1Rep1.18.bam
</span></span><span class="line"><span class="cl">samtools index H1Rep1.18.sorted.bam
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul>
<p>3、使用seq命令删除hg19_Ensembl_chr.bed文件中染色体名称的chr前缀（如有必要）。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">sed <span class="s1">&#39;s/^chr//&#39;</span> hg19_Ensembl_chr.bed &gt;hg19_Ensembl.bed
</span></span></code></pre></td></tr></table>
</div>
</div><p>4、使用RSeQC中的read_distribution.py计算read对不同基因组特征类型的分布。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">python3 ~/software/RSeQC-5.0.1/scripts/read_distribution.py -r hg19_Ensembl.bed -i H1Rep1.18.sorted.bam
</span></span></code></pre></td></tr></table>
</div>
</div><p>结果将以表格形式报告read和tag的总数以及所属类型。</p>
<p>5、使用RSeQC中的geneBody_coverage.py可以生成一个覆盖图来检查转录本的覆盖是否均匀，即是否存在3’端或5’端偏差。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">python3 ~/software/RSeQC-5.0.1/scripts/geneBody_coverage.py -r hg19_Ensembl_chr.bed -i H1Rep1.18.bam -o file
</span></span></code></pre></td></tr></table>
</div>
</div><p>运行该程序需要R在$PATH上。</p>
<p>6、使用RPKM_saturation.py对read的子集重新取样，对每个子集按RPKM单位计算丰度，并检查他们是否稳定。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">python3 ~/software/RSeQC-5.0.1/scripts/RPKM_saturation.py -r hg19_Ensembl_chr.bed -i H1Rep1.18.bam -o file
</span></span></code></pre></td></tr></table>
</div>
</div><p>7、使用junction_annotation.py将剪接点划分为“新的”、“部分新”的和“已知的”，分别代表0个、1个、2个剪接位点在参考基因模型中。结果以饼图显示。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">python3 ~/software/RseQC-5.0.1/scripts/junction_annotation.py -r hg19_Ensembl_chr.bed -i H1Rep1.18.bam -o file3
</span></span></code></pre></td></tr></table>
</div>
</div><p>8、使用junction_saturation.py检查剪接点的测序饱和状态。软件将检测到的剪接点标记为新的、部分新的和已知的，并通过重新抽样分析其饱和状态。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">python3 ~/software/RSeQC-5.0.1/scripts/junction_saturation.py -r hg19_Ensembl_chr.bed -i H1Rep1.18.bam -o file4
</span></span></code></pre></td></tr></table>
</div>
</div><p>9、使用HTSeq-count输出每个基因的计数。HTSeq-count需要SAM/BAM格式的比对文件，以及一个GTF格式的基因组注释文件。Htseq-count期望双端数据按读段名称进行排序，同时BAM文件和注释文件应当有相同的基因名称，本例中需要将GTF文件中基因名称1、2&hellip;转化为chr1、chr2&hellip;</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">wget ftp://ftp.ensembl.org/pub/release-74/gtf/homo_sapiens/Homo_sapiens.GRCh37.74.gtf.gz
</span></span><span class="line"><span class="cl">gunzip Homo_sapiens.GRCh37.74.gtf.gz
</span></span><span class="line"><span class="cl">samtools sort -o H1Rep1.18.namesorted.bam H1Rep1.18.bam
</span></span><span class="line"><span class="cl">sed <span class="s1">&#39;s/^\([0-9XYM]\)/chr**\1**/&#39;</span> Homo_sapiens.GRCh37.74.gtf &gt; Homo_sapiens.GRCh37.74.chr.gtf
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no H1Rep1.18.namesorted.bam Homo_sapiens.GRCh37.74.chr.gtf &gt;counts.txt
</span></span></code></pre></td></tr></table>
</div>
</div><p>10、为了方便对差异进行统计检验，最好删除counts.txt的后5行。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl">head -n -5 counts.txt &gt; genecounts.txt
</span></span></code></pre></td></tr></table>
</div>
</div><p>至此，我们已经由原始的reads得到了每个基因的counts。下一步，我们将从counts往后进行分析。</p>
<h4 id="讨论-1">讨论
</h4><h5 id="讨论1如何度量rna-seq数据的质量">讨论1：如何度量RNA-Seq数据的质量？
</h5><p>RNA-Seq读段基质量问题，如低置信度碱基问题，已经可以在原始读段水平上检测出，并且用phred64编码的形式报告在fastq文件中。当读段被作图到一个参考基因组，并且它们的位置与注释相匹配时，则可以进一步度量读段的质量，包括：</p>
<ul>
<li>
<p>测序深度的饱和性：检查测序深度是否接近饱和。</p>
</li>
<li>
<p>不同基因组特征之间的读段分布：比如读段在外显子、内含子以及基因间区的分布情况；读段在5’UTR和3’UTR之间的分布情况；读段在蛋白质编码基因、假基因、rRNA、miRNA之间的分布情况等等。</p>
</li>
<li>
<p>沿转录本的覆盖均匀性：不同的实验方法可能引入不同的位置偏差，例如一个polyA尾捕获步骤可能导致读段主要来自转录本的3’端。</p>
</li>
</ul>
<h5 id="讨论2如何度量基因表达丰度">讨论2：如何度量基因表达丰度？
</h5>$$
\text{RPKM} = \frac{\text{numReads}}{\left(\frac{\text{geneExonLength}}{1000}\right) \times \left(\frac{\text{totalNumReads}}{1000000}\right)} 
$$<p>
即，认为一条基因的read数与基因长度和总read数成正比，故将其作为分母除来归一化。这样可以在同一样本中比较不同基因的表达差异。</p>
$$
\text{FPKM} = \frac{\text{numFragments}}{\left(\frac{\text{geneExonLength}}{1000}\right) \times \left(\frac{\text{totalNumFragments}}{1000000}\right)}
$$$$
TPM_i = \frac{\frac{N_i}{L_i} \times 10^6}{\sum_{k=1}^{n} \frac{N_k}{L_k}} = \frac{RPKM_i}{\sum_{k=1}^{n} RPKM_k} \times 10^6
$$<p>
其中Ni指匹配到第i个转录本的read数，Li是第i个转录本的长度。TPM相当于是将RPKM或FPKM归一化，使得在不同样本中可以比较相同基因的表达差异。</p>
<h5 id="讨论3操作中输出的countstxt的数据结构">讨论3：操作中输出的counts.txt的数据结构。
</h5><p>该文件的最后几行如下所示：</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt">1
</span><span class="lnt">2
</span><span class="lnt">3
</span><span class="lnt">4
</span><span class="lnt">5
</span><span class="lnt">6
</span><span class="lnt">7
</span><span class="lnt">8
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-txt" data-lang="txt"><span class="line"><span class="cl">ENSG00000273491	0
</span></span><span class="line"><span class="cl">ENSG00000273492	0
</span></span><span class="line"><span class="cl">ENSG00000273493	0
</span></span><span class="line"><span class="cl">__no_feature	77237
</span></span><span class="line"><span class="cl">__ambiguous	52848
</span></span><span class="line"><span class="cl">__too_low_aQual	0
</span></span><span class="line"><span class="cl">__not_aligned	0
</span></span><span class="line"><span class="cl">__alignment_not_unique	93417
</span></span></code></pre></td></tr></table>
</div>
</div><p>最后5行分别代表没有进行计数的原因和对应的读段数目：比对与任何基因均不重叠、其比对与多个基因重叠、对比质量低于阈值、完全没有比对以及比对到基因组中的多个位置。</p>
<h3 id="bulk-rna测序下游分析从counts开始">Bulk-RNA测序下游分析——从counts开始
</h3><h4 id="数据准备">数据准备
</h4><div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span><span class="lnt">56
</span><span class="lnt">57
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-bash" data-lang="bash"><span class="line"><span class="cl"><span class="c1">#下载3个来自GM12892细胞系的BAM与BAI文件，以及4个来自H1hesc细胞系的BAM文件：</span>
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep1V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep2V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep3V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep1V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep2V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep3V2.bam
</span></span><span class="line"><span class="cl">wget -c http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeCaltechRnaSeq/wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep4V2.bam
</span></span><span class="line"><span class="cl"><span class="c1">#再分别重命名为GM1/2/3、H1/2/3/4 .bam</span>
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep1V2.bam GM1.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep2V2.bam GM2.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqGm12892R2x75Il200AlignsRep3V2.bam GM3.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep1V2.bam H1.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep2V2.bam H2.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep3V2.bam H3.bam
</span></span><span class="line"><span class="cl">mv wgEncodeCaltechRnaSeqH1hescR2x75Il200AlignsRep4V2.bam H4.bam
</span></span><span class="line"><span class="cl"><span class="c1">#下载人类基因组的注释文件，并将染色体名称加上“chr”。</span>
</span></span><span class="line"><span class="cl">wget ftp://ftp.ensembl.org/pub/release-74/gtf/homo_sapiens/Homo_sapiens.GRCh37.74.gtf.gz
</span></span><span class="line"><span class="cl">sed <span class="s1">&#39;s/^\([0-9XYM]\)/chr\1/&#39;</span> Homo_sapiens.GRCh37.74.gtf  Homo_sapiens.GRCh37.74.chr.gtf
</span></span><span class="line"><span class="cl"><span class="c1">#将分别提取18号染色体的子集。</span>
</span></span><span class="line"><span class="cl">samtools sort -o GM1.sorted.bam GM1.bam  
</span></span><span class="line"><span class="cl">samtools index GM1.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o GM1.sorted.18.bam GM1.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o GM2.sorted.bam GM2.bam  
</span></span><span class="line"><span class="cl">samtools index GM2.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o GM2.sorted.18.bam GM2.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o GM3.sorted.bam GM3.bam  
</span></span><span class="line"><span class="cl">samtools index GM3.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o GM3.sorted.18.bam GM3.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o H1.sorted.bam H1.bam  
</span></span><span class="line"><span class="cl">samtools index H1.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o H1.sorted.18.bam H1.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o H2.sorted.bam H2.bam  
</span></span><span class="line"><span class="cl">samtools index H2.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o H2.sorted.18.bam H2.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o H3.sorted.bam H3.bam  
</span></span><span class="line"><span class="cl">samtools index H3.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o H3.sorted.18.bam H3.sorted.bam chr18
</span></span><span class="line"><span class="cl">samtools sort -o H4.sorted.bam H4.bam  
</span></span><span class="line"><span class="cl">samtools index H4.sorted.bam 
</span></span><span class="line"><span class="cl">samtools view -b -o H4.sorted.18.bam H4.sorted.bam chr18
</span></span><span class="line"><span class="cl"><span class="c1">#转化为counts</span>
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no GM1.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; GM1.txt
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no GM2.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; GM2.txt
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no GM3.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; GM3.txt
</span></span><span class="line"><span class="cl">ython3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no H1.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; H1.txt
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no H2.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; H2.txt
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no H3.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; H3.txt
</span></span><span class="line"><span class="cl">python3 ~/software/htseq-main/scripts/htseq-count -f bam --stranded no H4.sorted.18.bam Homo_sapiens.GRCh37.74.chr.gtf &gt; H4.txt
</span></span><span class="line"><span class="cl"><span class="c1">#去除每个txt的后5行以便后续进行分析。</span>
</span></span><span class="line"><span class="cl">head -n -5 GM1.txt &gt; GM1.count.txt
</span></span><span class="line"><span class="cl">head -n -5 GM2.txt &gt; GM2.count.txt
</span></span><span class="line"><span class="cl">head -n -5 GM3.txt &gt; GM3.count.txt
</span></span><span class="line"><span class="cl">head -n -5 H1.txt &gt; H1.count.txt
</span></span><span class="line"><span class="cl">head -n -5 H2.txt &gt; H2.count.txt
</span></span><span class="line"><span class="cl">head -n -5 H3.txt &gt; H3.count.txt
</span></span><span class="line"><span class="cl">head -n -5 H4.txt &gt; H4.count.txt
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="使用r语言进行分析">使用R语言进行分析
</h4><p>1、将7个样本的counts合并到一个表格，然后保存在本地。在本例中，7个txt文件在工作目录下的data文件夹里。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="n">samples</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;GM&#39;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">),</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#39;H&#39;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">first.sample</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="s">&#34;C:\\Users\\admin\\Desktop\\Rscripts\\data\\&#34;</span><span class="p">,</span> <span class="n">samples[1]</span><span class="p">,</span><span class="s">&#34;.count.txt&#34;</span><span class="p">),</span><span class="n">header</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">count.table</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">first.sample</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="kr">for</span><span class="p">(</span><span class="n">s</span> <span class="kr">in</span> <span class="n">samples[2</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">samples</span><span class="p">)</span><span class="n">]</span><span class="p">){</span>
</span></span><span class="line"><span class="cl">  <span class="n">fname</span> <span class="o">&lt;-</span> <span class="nf">paste0</span><span class="p">(</span><span class="s">&#34;C:\\Users\\admin\\Desktop\\Rscripts\\data\\&#34;</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="s">&#34;.count.txt&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="n">column</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> <span class="n">row.names</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="n">count.table</span> <span class="o">&lt;-</span> <span class="nf">cbind</span><span class="p">(</span><span class="n">count.table</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="n">column</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="nf">colnames</span><span class="p">(</span><span class="n">count.table</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">samples</span>
</span></span><span class="line"><span class="cl"><span class="nf">write.table</span><span class="p">(</span><span class="n">count.table</span><span class="p">,</span><span class="n">file</span> <span class="o">=</span> <span class="s">&#34;count_table_chr18.txt&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">,</span><span class="n">quote</span> <span class="o">=</span> <span class="bp">F</span><span class="p">)</span> <span class="c1">#save as txt file</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>2、用DESeq包进行差异分析，差异基因存储在sig.deseq变量中。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#34;DESeq2&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">d.raw</span> <span class="o">&lt;-</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&#34;count_table_chr18.txt&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;\t&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">d</span> <span class="o">&lt;-</span> <span class="n">d.raw</span><span class="nf">[rowSums</span><span class="p">(</span><span class="n">d.raw</span><span class="o">&gt;</span><span class="m">3</span><span class="p">)</span><span class="o">&gt;</span><span class="m">2</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">grp</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="s">&#34;hES&#34;</span><span class="p">,</span><span class="m">4</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">cData</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span><span class="n">celltype</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="n">grp</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="nf">rownames</span><span class="p">(</span><span class="n">cData</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">d.deseq</span> <span class="o">&lt;-</span> <span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span> <span class="o">=</span> <span class="n">d</span><span class="p">,</span><span class="n">colData</span> <span class="o">=</span> <span class="n">cData</span><span class="p">,</span><span class="n">design</span> <span class="o">=</span> <span class="o">~</span><span class="n">celltype</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">d.deseq</span> <span class="o">&lt;-</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">d.deseq</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">results</span><span class="p">(</span><span class="n">d.deseq</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">res</span> <span class="o">&lt;-</span> <span class="nf">results</span><span class="p">(</span><span class="n">d.deseq</span><span class="p">,</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;celltype&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">sig</span> <span class="o">&lt;-</span> <span class="n">res</span><span class="nf">[which</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span> <span class="o">&lt;</span> <span class="m">0.01</span><span class="p">),</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">sig.deseq</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">sig</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>3、除此之外，还可以用不同的包进行分析，最后将几种方法得出的差异基因取交集。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="c1">#limma</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">limma</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;limma&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">limma</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">grp</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">des</span> <span class="o">&lt;-</span> <span class="nf">model.matrix</span><span class="p">(</span><span class="o">~</span><span class="m">0</span><span class="o">+</span><span class="n">grp</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">colnames</span><span class="p">(</span><span class="n">des</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;hES&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">contrast.matrix</span> <span class="o">&lt;-</span> <span class="nf">makeContrasts</span><span class="p">(</span><span class="n">GM</span><span class="o">-</span><span class="n">hES</span><span class="p">,</span><span class="n">levels</span><span class="o">=</span><span class="n">des</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">d.norm</span> <span class="o">&lt;-</span> <span class="nf">voom</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="n">design</span><span class="o">=</span><span class="n">des</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">fit</span> <span class="o">&lt;-</span> <span class="nf">lmFit</span><span class="p">(</span><span class="n">d.norm</span><span class="p">,</span><span class="n">des</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">contrasts.fit</span><span class="p">(</span><span class="n">fit</span><span class="p">,</span><span class="n">contrast.matrix</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">fit2</span> <span class="o">&lt;-</span> <span class="nf">eBayes</span><span class="p">(</span><span class="n">fit2</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span><span class="n">adjust</span><span class="o">=</span><span class="s">&#34;BH&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">all</span> <span class="o">&lt;-</span> <span class="nf">topTable</span><span class="p">(</span><span class="n">fit2</span><span class="p">,</span><span class="n">adjust.method</span><span class="o">=</span><span class="s">&#34;BH&#34;</span><span class="p">,</span><span class="n">number</span><span class="o">=</span><span class="m">10000</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">sig</span> <span class="o">&lt;-</span> <span class="n">all[all</span><span class="o">$</span><span class="n">adj.P.Val</span> <span class="o">&lt;</span> <span class="m">1e-2</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">sig.limma</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">sig</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">limma</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">sig</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#samr</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">samr</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;samr&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">impute</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#34;impute&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">samr</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">num.grp</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">4</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">samfit</span> <span class="o">&lt;-</span> <span class="nf">SAMseq</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="n">num.grp</span><span class="p">,</span><span class="n">resp.type</span><span class="o">=</span><span class="s">&#34;Two class unpaired&#34;</span><span class="p">,</span><span class="n">genenames</span><span class="o">=</span><span class="nf">rownames</span><span class="p">(</span><span class="n">d</span><span class="p">),</span><span class="n">random.seed</span><span class="o">=</span><span class="m">101010</span><span class="p">,</span><span class="n">fdr.output</span><span class="o">=</span><span class="m">0.01</span><span class="p">,</span><span class="n">nperms</span><span class="o">=</span><span class="m">1000</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">sig.sam</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="n">samfit</span><span class="o">$</span><span class="n">siggenes.table</span><span class="o">$</span><span class="n">genes.up[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">,</span><span class="n">samfit</span><span class="o">$</span><span class="n">siggenes.table</span><span class="o">$</span><span class="n">genes.lo[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1">#edgeR</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">edgeR</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;edgeR&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">edgeR</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.dgelist</span> <span class="o">=</span> <span class="nf">DGEList</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">d</span><span class="p">,</span><span class="n">group</span><span class="o">=</span><span class="nf">factor</span><span class="p">(</span><span class="n">grp</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.dgelist</span> <span class="o">=</span> <span class="nf">calcNormFactors</span><span class="p">(</span><span class="n">edgeR.dgelist</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s">&#34;TMM&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.dgelist</span> <span class="o">=</span> <span class="nf">estimateCommonDisp</span><span class="p">(</span><span class="n">edgeR.dgelist</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.dgelist</span> <span class="o">=</span> <span class="nf">estimateTagwiseDisp</span><span class="p">(</span><span class="n">edgeR.dgelist</span><span class="p">,</span><span class="n">trend</span> <span class="o">=</span> <span class="s">&#34;movingave&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.test</span> <span class="o">=</span> <span class="nf">exactTest</span><span class="p">(</span><span class="n">edgeR.dgelist</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.pvalues</span> <span class="o">=</span> <span class="n">edgeR.test</span><span class="o">$</span><span class="n">table</span><span class="o">$</span><span class="n">PValue</span>
</span></span><span class="line"><span class="cl"><span class="n">edgeR.adjpvalues</span> <span class="o">=</span> <span class="nf">p.adjust</span><span class="p">(</span><span class="n">edgeR.pvalues</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s">&#34;BH&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">sig.table</span> <span class="o">&lt;-</span> <span class="n">edgeR.test</span><span class="o">$</span><span class="n">table</span><span class="nf">[which</span><span class="p">(</span><span class="n">edgeR.adjpvalues</span> <span class="o">&lt;</span><span class="m">0.01</span><span class="p">),</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">sig.edgeR</span> <span class="o">&lt;-</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">sig.table</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>4、进行可视化处理，包括：进行主成分分析判断两组数据有何差异；对四种分析方法分析出的差异基因做Venn图并找出共有的差异基因；绘制热图可视化基因表达；绘制火山图进一步筛选基因。</p>
<ul>
<li>
<p>主成分分析</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="c1"># PCA GENE fig.A1</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">ggbiplot</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;ggbiplot&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">ggbiplot</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">ct</span> <span class="o">&lt;-</span> <span class="n">count.table</span><span class="nf">[rownames</span><span class="p">(</span><span class="n">count.table</span><span class="p">)</span> <span class="o">%in%</span> <span class="n">sig.deseq</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">ct</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="nf">scale</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">ct</span><span class="p">)))</span>
</span></span><span class="line"><span class="cl"><span class="n">ct.pca</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="n">ct</span><span class="p">,</span><span class="n">scale.</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">ggbiplot</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">ct.pca</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">choices</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">  <span class="n">obs.scale</span><span class="o">=</span><span class="m">1</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">var.scale</span><span class="o">=</span><span class="m">1</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">var.axes</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">varname.size</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">varname.adjust</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme_bw</span><span class="p">()</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.direction</span> <span class="o">=</span> <span class="s">&#39;horizontal&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.position</span> <span class="o">=</span> <span class="s">&#39;top&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">ggsave</span><span class="p">(</span><span class="s">&#39;pca_gene.pdf&#39;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">family</span><span class="o">=</span><span class="s">&#34;GB1&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># PCA TYPE fig.A2</span>
</span></span><span class="line"><span class="cl"><span class="n">coln</span> <span class="o">=</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">rown</span> <span class="o">=</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">ct</span><span class="o">=</span><span class="nf">t</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>  <span class="c1">#Transpose</span>
</span></span><span class="line"><span class="cl"><span class="nf">colnames</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span><span class="o">=</span><span class="n">rown</span>
</span></span><span class="line"><span class="cl"><span class="nf">rownames</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span><span class="o">=</span><span class="n">coln</span>
</span></span><span class="line"><span class="cl"><span class="n">ct</span> <span class="o">&lt;-</span> <span class="n">ct[</span><span class="p">,</span> <span class="nf">colSums</span><span class="p">(</span><span class="n">ct</span> <span class="o">!=</span> <span class="m">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="n">]</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">ct.pca</span> <span class="o">&lt;-</span> <span class="nf">prcomp</span><span class="p">(</span><span class="n">ct</span><span class="p">,</span><span class="n">scale.</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">ggbiplot</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">ct.pca</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">choices</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">  <span class="n">obs.scale</span><span class="o">=</span><span class="m">1</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">var.scale</span><span class="o">=</span><span class="m">1</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">var.axes</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">labels</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">ct</span><span class="p">)),</span>
</span></span><span class="line"><span class="cl">  <span class="n">groups</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">)</span> <span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">ellipse</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">ellipse.prob</span> <span class="o">=</span> <span class="m">0.95</span><span class="p">,</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&#39;#70abd8&#39;</span><span class="p">,</span><span class="s">&#39;#fb9b8e&#39;</span><span class="p">))</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme_bw</span><span class="p">()</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.direction</span> <span class="o">=</span> <span class="s">&#39;horizontal&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.position</span> <span class="o">=</span> <span class="s">&#39;top&#39;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.text</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">    <span class="n">legend.title</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">ggsave</span><span class="p">(</span><span class="s">&#39;pca_type.pdf&#39;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">family</span><span class="o">=</span><span class="s">&#34;GB1&#34;</span><span class="p">)</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>画出四组差异基因的Venn图。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="c1"># Venn Graph fig.B</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">VennDiagram</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#34;VennDiagram&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">VennDiagram</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">gene.list</span> <span class="o">=</span> <span class="nf">list</span><span class="p">(</span><span class="n">sig.deseq</span><span class="p">,</span><span class="n">sig.limma</span><span class="p">,</span><span class="n">sig.edgeR</span><span class="p">,</span><span class="n">sig.sam</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">venn1</span> <span class="o">&lt;-</span> <span class="nf">venn.diagram</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">gene.list</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">category.names</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;deseq&#34;</span><span class="p">,</span><span class="s">&#34;limma&#34;</span><span class="p">,</span><span class="s">&#34;edgeR&#34;</span><span class="p">,</span><span class="s">&#34;sam&#34;</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">  <span class="n">alpha</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">cat.cex</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">cat.fontface</span> <span class="o">=</span> <span class="s">&#34;bold&#34;</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">lty</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">cat.pos</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">filename</span> <span class="o">=</span> <span class="kc">NULL</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">resolution</span> <span class="o">=</span> <span class="m">300</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">fill</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;red&#34;</span><span class="p">,</span> <span class="s">&#34;blue&#34;</span><span class="p">,</span> <span class="s">&#34;green&#34;</span><span class="p">,</span> <span class="s">&#34;yellow&#34;</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">  <span class="n">compression</span> <span class="o">=</span> <span class="s">&#34;lzw&#34;</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;venn.pdf&#34;</span><span class="p">,</span><span class="n">width</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span><span class="n">height</span> <span class="o">=</span> <span class="m">10</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">grid.draw</span><span class="p">(</span><span class="n">venn1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">dev.off</span><span class="p">()</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>对差异基因绘制热图</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="c1"># Pheat map fig.C1</span>
</span></span><span class="line"><span class="cl"><span class="kr">if</span> <span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">install.packages</span><span class="p">(</span><span class="s">&#34;pheatmap&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">save_pheatmap_pdf</span> <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">7</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="m">7</span><span class="p">)</span> <span class="p">{</span>
</span></span><span class="line"><span class="cl">  <span class="nf">stopifnot</span><span class="p">(</span><span class="o">!</span><span class="nf">missing</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">stopifnot</span><span class="p">(</span><span class="o">!</span><span class="nf">missing</span><span class="p">(</span><span class="n">filename</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">  <span class="nf">pdf</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">width</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="n">height</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="n">grid</span><span class="o">::</span><span class="nf">grid.newpage</span><span class="p">()</span>
</span></span><span class="line"><span class="cl">  <span class="n">grid</span><span class="o">::</span><span class="nf">grid.draw</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">gtable</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">  <span class="nf">dev.off</span><span class="p">()</span>
</span></span><span class="line"><span class="cl"><span class="p">}</span>
</span></span><span class="line"><span class="cl"><span class="n">common_diff_genes</span> <span class="o">&lt;-</span> <span class="nf">Reduce</span><span class="p">(</span><span class="n">intersect</span><span class="p">,</span> <span class="nf">list</span><span class="p">(</span><span class="n">sig.deseq</span><span class="p">,</span><span class="n">sig.edgeR</span><span class="p">,</span><span class="n">sig.limma</span><span class="p">,</span><span class="n">sig.sam</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">ct</span> <span class="o">&lt;-</span>   <span class="n">count.table</span><span class="nf">[rownames</span><span class="p">(</span><span class="n">count.table</span><span class="p">)</span> <span class="o">%in%</span>  <span class="n">common_diff_genes</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">ct</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="nf">scale</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">ct</span><span class="p">)))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">annotation_col</span> <span class="o">&lt;-</span> <span class="nf">data.frame</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">celltype</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;GM&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">,</span><span class="s">&#34;H1&#34;</span><span class="p">)</span> 
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">rownames</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">ct</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">pm</span> <span class="o">&lt;-</span> <span class="nf">pheatmap</span><span class="p">(</span>
</span></span><span class="line"><span class="cl">  <span class="n">ct</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">fontsize</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span>                
</span></span><span class="line"><span class="cl">  <span class="n">fontsize_row</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>             
</span></span><span class="line"><span class="cl">  <span class="n">fontsize_col</span> <span class="o">=</span> <span class="m">2</span><span class="p">,</span>  
</span></span><span class="line"><span class="cl">  <span class="n">display_numbers</span> <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">annotation_col</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">  <span class="n">color</span> <span class="o">=</span> <span class="nf">colorRampPalette</span><span class="p">(</span><span class="n">colors</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;blue&#34;</span><span class="p">,</span><span class="s">&#34;white&#34;</span><span class="p">,</span><span class="s">&#34;red&#34;</span><span class="p">))(</span><span class="m">100</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">save_pheatmap_pdf</span><span class="p">(</span><span class="n">pm</span><span class="p">,</span> <span class="s">&#34;C:\\Users\\admin\\Desktop\\Rscripts\\pheatmap.pdf&#34;</span><span class="p">,</span><span class="m">6</span><span class="p">,</span><span class="m">12</span><span class="p">)</span> 
</span></span><span class="line"><span class="cl"><span class="c1"># Pheat map fig.C2</span>
</span></span><span class="line"><span class="cl"><span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;hm2.pdf&#34;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="m">5</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">heatmap</span><span class="p">(</span><span class="nf">cor</span><span class="p">(</span><span class="n">ct</span><span class="p">),</span><span class="n">cexCol</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span><span class="n">cexRow</span><span class="o">=</span><span class="m">0.75</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">dev.off</span><span class="p">()</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>绘制火山图、可视化差异基因的log2FC和P值。</p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre tabindex="0" class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span></code></pre></td>
<td class="lntd">
<pre tabindex="0" class="chroma"><code class="language-R" data-lang="R"><span class="line"><span class="cl"><span class="n">expr_matrix</span> <span class="o">&lt;-</span> <span class="n">count.table</span><span class="nf">[rownames</span><span class="p">(</span><span class="n">count.table</span><span class="p">)</span> <span class="o">%in%</span> <span class="n">common_diff_genes</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">GM_samples</span> <span class="o">&lt;-</span> <span class="m">1</span><span class="o">:</span><span class="m">3</span>   <span class="c1"># 前三列为 GM</span>
</span></span><span class="line"><span class="cl"><span class="n">H1_samples</span> <span class="o">&lt;-</span> <span class="m">4</span><span class="o">:</span><span class="m">7</span>   <span class="c1"># 后四列为 H1</span>
</span></span><span class="line"><span class="cl"><span class="n">sig2</span> <span class="o">&lt;-</span> <span class="n">sig</span><span class="nf">[rownames</span><span class="p">(</span><span class="n">sig</span><span class="p">)</span> <span class="o">%in%</span> <span class="n">common_diff_genes</span><span class="p">,</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="n">Gene_table</span> <span class="o">=</span> <span class="nf">data.frame</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">sig2</span><span class="p">),</span><span class="n">sig2</span><span class="o">$</span><span class="n">logFC</span><span class="p">,</span><span class="n">sig2</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">rownames</span><span class="p">(</span><span class="n">Gene_table</span><span class="p">)</span> <span class="o">=</span> <span class="n">Gene_table</span><span class="o">$</span><span class="n">rownames.sig2.</span>
</span></span><span class="line"><span class="cl"><span class="n">Gene_table</span> <span class="o">=</span> <span class="n">Gene_table[</span><span class="p">,</span><span class="m">-1</span><span class="n">]</span>
</span></span><span class="line"><span class="cl"><span class="nf">colnames</span><span class="p">(</span><span class="n">Gene_table</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;log2FC&#34;</span><span class="p">,</span><span class="s">&#34;pvalue&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">change</span> <span class="o">=</span> <span class="nf">as.factor</span><span class="p">(</span><span class="nf">ifelse</span><span class="p">(</span><span class="n">Gene_table</span><span class="o">$</span><span class="n">pvalue</span> <span class="o">&lt;</span> <span class="m">0.05</span> <span class="o">&amp;</span> <span class="nf">abs</span><span class="p">(</span><span class="n">Gene_table</span><span class="o">$</span><span class="n">log2FC</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">1</span><span class="p">,</span>
</span></span><span class="line"><span class="cl">                          <span class="nf">ifelse</span><span class="p">(</span><span class="n">Gene_table</span><span class="o">$</span><span class="n">log2FC</span> <span class="o">&gt;</span> <span class="m">1</span> <span class="p">,</span><span class="s">&#39;Up&#39;</span><span class="p">,</span><span class="s">&#39;Down&#39;</span><span class="p">),</span><span class="s">&#39;No change&#39;</span><span class="p">))</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="nf">pdf</span><span class="p">(</span><span class="s">&#34;vol.pdf&#34;</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="m">5</span><span class="p">,</span><span class="n">height</span><span class="o">=</span><span class="m">5</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">ggplot</span><span class="p">(</span><span class="n">Gene_table</span><span class="p">,</span><span class="nf">aes</span><span class="p">(</span><span class="n">log2FC</span><span class="p">,</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)))</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">geom_hline</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="o">-</span><span class="nf">log10</span><span class="p">(</span><span class="m">0.05</span><span class="p">),</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s">&#34;dashed&#34;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&#34;#999999&#34;</span><span class="p">)</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">-1</span><span class="p">,</span><span class="m">1</span><span class="p">),</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s">&#34;dashed&#34;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&#34;#999999&#34;</span><span class="p">)</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">geom_point</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">color</span> <span class="o">=</span> <span class="n">change</span><span class="p">),</span>
</span></span><span class="line"><span class="cl">             <span class="n">size</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">,</span> 
</span></span><span class="line"><span class="cl">             <span class="n">alpha</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">)</span> <span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme_bw</span><span class="p">(</span><span class="n">base_size</span> <span class="o">=</span> <span class="m">12</span><span class="p">)</span><span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_color_jama</span><span class="p">()</span> <span class="o">+</span>
</span></span><span class="line"><span class="cl">  <span class="nf">theme</span><span class="p">(</span><span class="n">panel.grid</span> <span class="o">=</span> <span class="nf">element_blank</span><span class="p">(),</span>
</span></span><span class="line"><span class="cl">        <span class="n">legend.position</span> <span class="o">=</span> <span class="s">&#39;right&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">dev.off</span><span class="p">()</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul>
<h5 id="图片">图片
</h5><p><img src="/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/fig.png"
	width="1191"
	height="1684"
	srcset="/p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/fig_hu13588729010768890305.png 480w, /p/rna%E6%B5%8B%E5%BA%8F%E5%88%86%E6%9E%90%E7%9A%84%E6%A0%87%E5%87%86%E6%B5%81%E7%A8%8B/fig_hu8964382468325597058.png 1024w"
	loading="lazy"
	
		alt="image-20241214194821212"
	
	
		class="gallery-image" 
		data-flex-grow="70"
		data-flex-basis="169px"
	
></p>

</section>


    <footer class="article-footer">
    

    
    <section class="article-copyright">
        <svg xmlns="http://www.w3.org/2000/svg" class="icon icon-tabler icon-tabler-copyright" width="24" height="24" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
  <path stroke="none" d="M0 0h24v24H0z"/>
  <circle cx="12" cy="12" r="9" />
  <path d="M14.5 9a3.5 4 0 1 0 0 6" />
</svg>



        <span>Licensed under CC BY-NC-SA 4.0</span>
    </section>
    </footer>


    
        <link 
                rel="stylesheet" 
                href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"integrity="sha384-n8MVd4RsNIU0tAv4ct0nTaAbDJwPJzDEaqSD1odI&#43;WdtXRGWt2kTvGFasHpSy3SV"crossorigin="anonymous"
            ><script 
                src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.js"integrity="sha384-XjKyOOlGwcjNTAIQHIpgOno0Hl1YQqzUOEleOLALmuqehneUG&#43;vnGctmUb0ZY0l8"crossorigin="anonymous"
                defer
                >
            </script><script 
                src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/auto-render.min.js"integrity="sha384-&#43;VBxd3r6XgURycqtZ117nYw44OOcIax56Z4dCRWbxyPt0Koah1uHoK0o4&#43;/RRE05"crossorigin="anonymous"
                defer
                >
            </script><script>
    window.addEventListener("DOMContentLoaded", () => {
        renderMathInElement(document.body, {
            delimiters: [
                { left: "$$", right: "$$", display: true },
                { left: "$", right: "$", display: false },
                { left: "\\(", right: "\\)", display: false },
                { left: "\\[", right: "\\]", display: true }
            ],
            ignoredClasses: ["gist"]
        });})
</script>
    
</article>

    

    

     
    
        
    

    <footer class="site-footer">
    <section class="copyright">
        &copy; 
        
        2024 鳕鱼
    </section>
    
    <section class="powerby">
        
            welcome to my blog <br/>
        使用 <a href="https://gohugo.io/" target="_blank" rel="noopener">Hugo</a> 构建 <br />
        主题 <b><a href="https://github.com/CaiJimmy/hugo-theme-stack" target="_blank" rel="noopener" data-version="3.29.0">Stack</a></b> 由 <a href="https://jimmycai.com" target="_blank" rel="noopener">Jimmy</a> 设计
    </section>
</footer>


    
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

    
    <div class="pswp__bg"></div>

    
    <div class="pswp__scroll-wrap">

        
        <div class="pswp__container">
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
        </div>

        
        <div class="pswp__ui pswp__ui--hidden">

            <div class="pswp__top-bar">

                

                <div class="pswp__counter"></div>

                <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>

                <button class="pswp__button pswp__button--share" title="Share"></button>

                <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>

                <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>

                
                
                <div class="pswp__preloader">
                    <div class="pswp__preloader__icn">
                        <div class="pswp__preloader__cut">
                            <div class="pswp__preloader__donut"></div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
                <div class="pswp__share-tooltip"></div>
            </div>

            <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
            </button>

            <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
            </button>

            <div class="pswp__caption">
                <div class="pswp__caption__center"></div>
            </div>

        </div>

    </div>

</div><script 
                src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.js"integrity="sha256-ePwmChbbvXbsO02lbM3HoHbSHTHFAeChekF1xKJdleo="crossorigin="anonymous"
                defer
                >
            </script><script 
                src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js"integrity="sha256-UKkzOn/w1mBxRmLLGrSeyB4e1xbrp4xylgAWb3M42pU="crossorigin="anonymous"
                defer
                >
            </script><link 
                rel="stylesheet" 
                href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/default-skin/default-skin.min.css"crossorigin="anonymous"
            ><link 
                rel="stylesheet" 
                href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.css"crossorigin="anonymous"
            >

            </main>
        </div>
        <script 
                src="https://cdn.jsdelivr.net/npm/node-vibrant@3.1.6/dist/vibrant.min.js"integrity="sha256-awcR2jno4kI5X0zL8ex0vi2z&#43;KMkF24hUW8WePSA9HM="crossorigin="anonymous"
                
                >
            </script><script type="text/javascript" src="/ts/main.1e9a3bafd846ced4c345d084b355fb8c7bae75701c338f8a1f8a82c780137826.js" defer></script>
<script>
    (function () {
        const customFont = document.createElement('link');
        customFont.href = "https://fonts.googleapis.com/css2?family=Lato:wght@300;400;700&display=swap";

        customFont.type = "text/css";
        customFont.rel = "stylesheet";

        document.head.appendChild(customFont);
    }());
</script>
<script src=https://Detarame-ding.github.io/background/sakura.js></script>

    </body>
</html>
